File: 2024103410刘序桐/项目/python代码/maxHeap.py
class MHNode:
    """最大堆节点类"""
    def __init__(self, i=0, j=0, delta=0.0):
        self.i = i          # 合并的簇id
        self.j = j          # 合并的簇id
        self.index = 0      # 该簇在最大堆中的索引
        self.delta = delta  # 最大堆排序依据


class MaxHeap:
    """最大堆类"""
    
    def __init__(self):
        self.heap = []  # 动态数组，存储MHNode对象
    
    def parent(self, index):
        """获取父节点索引"""
        return (index - 1) // 2
    
    def left_child(self, index):
        """获取左子节点索引"""
        return 2 * index + 1
    
    def right_child(self, index):
        """获取右子节点索引"""
        return 2 * index + 2
    
    def heapify_up(self, index):
        """堆化向上"""
        while index > 0 and self.heap[self.parent(index)].delta < self.heap[index].delta:
            parent_idx = self.parent(index)
            # 交换节点
            self.heap[parent_idx], self.heap[index] = self.heap[index], self.heap[parent_idx]
            # 更新索引
            self.heap[parent_idx].index = parent_idx
            self.heap[index].index = index
            index = parent_idx
        self.heap[index].index = index  # 确保最终位置的索引正确
    
    def heapify_down(self, index):
        """堆化向下"""
        largest = index
        left = self.left_child(index)
        right = self.right_child(index)
        
        if left < len(self.heap) and self.heap[left].delta > self.heap[largest].delta:
            largest = left
        
        if right < len(self.heap) and self.heap[right].delta > self.heap[largest].delta:
            largest = right
        
        if largest != index:
            # 交换节点
            self.heap[index], self.heap[largest] = self.heap[largest], self.heap[index]
            # 更新索引
            self.heap[index].index = index
            self.heap[largest].index = largest
            self.heapify_down(largest)
    
    def insert(self, node):
        """插入元素"""
        self.heap.append(node)
        node.index = len(self.heap) - 1  # 设置新插入节点的索引
        self.heapify_up(len(self.heap) - 1)
    
    def get_max(self):
        """获取堆顶元素"""
        if not self.heap:
            raise RuntimeError("Heap is empty")
        return self.heap[0]
    
    def remove_max(self):
        """删除堆顶元素"""
        if not self.heap:
            raise RuntimeError("Heap is empty")
        
        # 将最后一个元素放到堆顶
        self.heap[0] = self.heap[-1]
        self.heap[0].index = 0  # 更新堆顶元素的新索引
        self.heap.pop()
        
        if self.heap:
            self.heapify_down(0)
    
    def size(self):
        """获取堆的大小"""
        return len(self.heap)
    
    def is_empty(self):
        """检查堆是否为空"""
        return len(self.heap) == 0
    
    def remove_at_index(self, index):
        """根据索引删除元素"""
        if index < 0 or index >= len(self.heap):
            raise RuntimeError("Index out of bounds")
        
        # 将要删除的元素与最后一个元素交换
        self.heap[index], self.heap[-1] = self.heap[-1], self.heap[index]
        self.heap[index].index = index  # 更新交换后节点的索引
        self.heap.pop()  # 删除最后一个元素
        
        # 恢复堆的性质
        if index < len(self.heap):
            self.heapify_down(index)  # 尝试向下堆化
            self.heapify_up(index)    # 如果向下堆化无效，尝试向上堆化
    
    def delete_node(self, index):
        """删除指定节点，并更新树结构"""
        if index < 0 or index >= len(self.heap):
            raise RuntimeError("Index out of bounds")
        
        # 将索引和最后一个元素互换
        self.heap[index], self.heap[-1] = self.heap[-1], self.heap[index]
        self.heap[index].index = index  # 更新交换后节点的索引
        self.heap.pop()  # 删除最后一个元素
        
        if index < len(self.heap):
            self.heapify_down(index)  # 尝试向下堆化
            self.heapify_up(index)    # 如果向下堆化无效，尝试向上堆化
    
    def print_heap(self):
        """输出整个最大堆的内容"""
        for node in self.heap:
            print(f"{node.delta} ", end="")
        print()


# 示例使用
if __name__ == "__main__":
    # 创建最大堆
    max_heap = MaxHeap()
    
    # 创建一些节点并插入
    node1 = MHNode(1, 2, 10.5)
    node2 = MHNode(3, 4, 15.2)
    node3 = MHNode(5, 6, 8.7)
    node4 = MHNode(7, 8, 20.1)
    
    max_heap.insert(node1)
    max_heap.insert(node2)
    max_heap.insert(node3)
    max_heap.insert(node4)
    
    print("堆内容:")
    max_heap.print_heap()
    
    print(f"堆大小: {max_heap.size()}")
    print(f"最大元素的delta值: {max_heap.get_max().delta}")
    
    # 删除最大元素
    max_heap.remove_max()
    print("删除最大元素后:")
    max_heap.print_heap()

File: 2024103410刘序桐/项目/C++代码/maxHeap.cpp
#include "maxHeap.h"

// 获取父节点索引
int MaxHeap::parent(int index)
{
    return (index - 1) / 2;
}

// 获取左子节点索引
int MaxHeap::leftChild(int index)
{
    return 2 * index + 1;
}

// 获取右子节点索引
int MaxHeap::rightChild(int index)
{
    return 2 * index + 2;
}

// 堆化向上
void MaxHeap::heapifyUp(int index)
{
    while (index > 0 && heap[parent(index)]->delta < heap[index]->delta)
    {
        std::swap(heap[parent(index)], heap[index]);
        heap[parent(index)]->index = parent(index); // 更新父节点的索引
        heap[index]->index = index;                 // 更新当前节点的索引
        index = parent(index);
    }
    heap[index]->index = index; // 确保最终位置的索引正确
}

// 堆化向下
void MaxHeap::heapifyDown(int index)
{
    int largest = index;
    int left = leftChild(index);
    int right = rightChild(index);

    if (left < heap.size() && heap[left]->delta > heap[largest]->delta)
    {
        largest = left;
    }

    if (right < heap.size() && heap[right]->delta > heap[largest]->delta)
    {
        largest = right;
    }

    if (largest != index)
    {
        std::swap(heap[index], heap[largest]);
        heap[index]->index = index;     // 更新当前节点的索引
        heap[largest]->index = largest; // 更新交换后节点的索引
        heapifyDown(largest);
    }
}

// 插入元素
void MaxHeap::insert(MHNode *node)
{
    heap.push_back(node);
    node->index = heap.size() - 1; // 设置新插入节点的索引
    heapifyUp(heap.size() - 1);
}

// 获取堆顶元素
MHNode *MaxHeap::getMax()
{
    if (heap.empty())
    {
        throw std::runtime_error("Heap is empty");
    }
    return heap[0];
}

// 删除堆顶元素
void MaxHeap::removeMax()
{
    if (heap.empty())
    {
        throw std::runtime_error("Heap is empty");
    }
    heap[0] = heap.back(); // 将最后一个元素放到堆顶
    heap[0]->index = 0;    // 更新堆顶元素的新索引
    heap.pop_back();
    if (!heap.empty())
    {
        heapifyDown(0);
    }
}

// 获取堆的大小
int MaxHeap::size() const
{
    return heap.size();
}

// 检查堆是否为空
bool MaxHeap::isEmpty() const
{
    return heap.empty();
}

// 根据索引删除元素
void MaxHeap::removeAtIndex(int index)
{
    if (index < 0 || index >= heap.size())
    {
        throw std::runtime_error("Index out of bounds");
    }

    // 将要删除的元素与最后一个元素交换
    std::swap(heap[index], heap.back());
    heap[index]->index = index; // 更新交换后节点的索引
    heap.pop_back();            // 删除最后一个元素

    // 恢复堆的性质
    if (index < heap.size())
    {
        heapifyDown(index); // 尝试向下堆化
        heapifyUp(index);   // 如果向下堆化无效，尝试向上堆化
    }
}

// 输出整个最大堆的内容
void MaxHeap::printHeap() const
{
    // std::cout << "MaxHeap contents:" << std::endl;
    for (const auto &node : heap)
    {
        /*std::cout << "Node(i: " << node->i
                  << ", j: " << node->j
                  << ", index: " << node->index
                  << ", delta: " << node->delta
                  << ")" << std::endl;*/
        std::cout << node->delta << " ";
    }
    std::cout << std::endl;
}

void MaxHeap::deleteNode(int index)
{
    // 根据索引删除元素
    // 将索引和最后一个元素互换
    // 然后删除最后一个元素,然后将换完的元素向下堆化
    // 如果向下堆化无效，尝试向上堆化
    if (index < 0 || index >= heap.size())
    {
        throw std::runtime_error("Index out of bounds");
    }
    std::swap(heap[index], heap.back());
    heap[index]->index = index; // 更新交换后节点的索引
    heap.pop_back();            // 删除最后一个元素
    if (index < heap.size())
    {
        heapifyDown(index); // 尝试向下堆化
        heapifyUp(index);   // 如果向下堆化无效，尝试向上堆化
    }

    // std::cout << "delte node success!" << std::endl;
}